### PR TITLE
bugfix: make ringbuffer right

### PR DESCRIPTION
**Pull request by fuweid**
Signed-off-by: Wei Fu <fhfuwei@163.com>

### Ⅰ. Describe what this PR did

Make ringbuffer right

### Ⅱ. Does this pull request fix one issue?

NONE

### Ⅲ. Describe how you did it

As the original design says, the ringbuffer uses two pointer to represent
push and pop position respectively in the circle list. The ringbuffer
allows to override oldest data if the buffer is full.

However, the Close action will be hang if the push pointer doesn't equal to
pop pointer. Please check the case.

```
E = empty, H = has data

At this time, the circle has full buffer, which means each node contains
the data.

   0H -> 1H -> 2H -> 3H -> 4H -> 0H
               ^     ^
               |     |
               |   push pointer
              pop pointer

In the following goroutine scheduled, the pop action will consume all
the data.

   0E -> 1E -> 2E -> 3E -> 4E -> 0E
               ^     ^
               |     |
               |   push pointer
              pop pointer

However, if there are not enough data to make the push pointer equals to
the pop pointer, the Close action will be hang.
```

In order to make the ringbuffer right, use FIFO queue to as buffer. The
front pointer contains the oldest data. When the buffer is full and
there is new data coming, the ringbuffer can drop the front one and push
new data into the back.

However, the ringbuffer still contains the data after closed. The
containerio should drain all the data and write it into the backend.

### Ⅳ. Describe how to verify it

It should be closed successfully.
```
➜  /tmp cat ./run.sh
#!/usr/bin/env bash

for _ in $(seq 1 1000); do
  pouch run --rm reg.docker.alibaba-inc.com/base/busybox sh -c 'for i in $(seq 1 10000); do echo hello-$i; done;' | wc -l
done
```

### Ⅴ. Special notes for reviews

The ringbuffer means non-blocked IO, which loss the data if the consumer pops slowly.
Next action:
* refactor containerio
* make blocked IO if necessary

### CODE REVIEW

**Comment by allencloud** [comment]
CRI tools validation fails:
```
[k8s.io] Container runtime should support basic operations on container 
  runtime should support execSync [Conformance]
  /home/travis/gopath/src/github.com/kubernetes-incubator/cri-tools/pkg/validate/container.go:121
[BeforeEach] [k8s.io] Container
  /home/travis/gopath/src/github.com/kubernetes-incubator/cri-tools/pkg/framework/framework.go:50
[BeforeEach] [k8s.io] Container
  /home/travis/gopath/src/github.com/kubernetes-incubator/cri-tools/pkg/validate/container.go:63
[BeforeEach] runtime should support basic operations on container
  /home/travis/gopath/src/github.com/kubernetes-incubator/cri-tools/pkg/validate/container.go:72
[It] runtime should support execSync [Conformance]
  /home/travis/gopath/src/github.com/kubernetes-incubator/cri-tools/pkg/validate/container.go:121
STEP: create container
STEP: Get image status for image: busybox:1.26
STEP: Create container.
Jun 20 13:09:09.131: INFO: Created container "517bed7d36812d99d5be947992a537f12d9b595710d374c7644da4942bd15fe7"
STEP: start container
STEP: Start container for containerID: 517bed7d36812d99d5be947992a537f12d9b595710d374c7644da4942bd15fe7
Jun 20 13:09:09.249: INFO: Started container "517bed7d36812d99d5be947992a537f12d9b595710d374c7644da4942bd15fe7"
STEP: test execSync
STEP: verify execSync output
STEP: execSync for containerID: 517bed7d36812d99d5be947992a537f12d9b595710d374c7644da4942bd15fe7
Jun 20 13:09:09.312: INFO: Execsync succeed
[AfterEach] runtime should support basic operations on container
  /home/travis/gopath/src/github.com/kubernetes-incubator/cri-tools/pkg/validate/container.go:76
STEP: stop PodSandbox
STEP: delete PodSandbox
[AfterEach] [k8s.io] Container
  /home/travis/gopath/src/github.com/kubernetes-incubator/cri-tools/pkg/framework/framework.go:51
• Failure [0.908 seconds]
[k8s.io] Container
/home/travis/gopath/src/github.com/kubernetes-incubator/cri-tools/pkg/framework/framework.go:72
  runtime should support basic operations on container
  /home/travis/gopath/src/github.com/kubernetes-incubator/cri-tools/pkg/validate/container.go:68
    runtime should support execSync [Conformance] [It]
    /home/travis/gopath/src/github.com/kubernetes-incubator/cri-tools/pkg/validate/container.go:121
    The stdout output of execSync should be hello
    
    Expected
        <string>: 
    to equal
        <string>: hello
        
    /home/travis/gopath/src/github.com/kubernetes-incubator/cri-tools/pkg/validate/container.go:380
```

ping @YaoZengzeng @Starnop 
And I retrigger this to have another try.

File: pkg/ringbuffer/list.go [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+func (q *queue) size() int {
+	return q.count
+}
+
+func (q *queue) enqueue(val interface{}) {
+	elem := elemPool.Get().(*element)
+	elem.val = val
+
+	at := q.root.prev
+	n := at.next
```
Comment (yyb196) [inline pkg/ringbuffer/list.go]: why we need this local var n, isn't it always q.root?

**Comment by YaoZengzeng** [comment]
LGTM, @yyb196 @HusterWan May review this PR more in detail 😄

File: pkg/ringbuffer/list.go (hunk 1 above)
Comment (fuweid) [reply pkg/ringbuffer/list.go]: I will make it clear.

**Comment by fuweid** [comment]
@yyb196 I have updated the code. Please take a look.

cc @YaoZengzeng

File: pkg/ringbuffer/list.go [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+
+func (q *queue) size() int {
+	return q.count
+}
+
+func (q *queue) enqueue(val interface{}) {
+	elem := elemPool.Get().(*element)
+	elem.val = val
+
+	at := q.root.prev
```
Comment (shaloulcy) [inline pkg/ringbuffer/list.go]: rename **at** to **last** or **tail**, it may more clear

File: pkg/ringbuffer/list.go [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+	q.root.prev = elem
+	q.count++
+}
+
+func (q *queue) dequeue() interface{} {
+	if q.size() == 0 {
+		return nil
+	}
+
+	at := q.root.next
```
Comment (shaloulcy) [inline pkg/ringbuffer/list.go]: rename **at** to **first** or **head**, it may more clear

File: pkg/ringbuffer/list.go (hunk 2 above)
Comment (fuweid) [reply pkg/ringbuffer/list.go]: Hi @shaloulcy , the `q.root.prev` means the tail or last. `at` means the position. I think we can keep it here.

File: pkg/ringbuffer/list.go (hunk 3 above)
Comment (fuweid) [reply pkg/ringbuffer/list.go]: Please check the last comment

**Comment by fuweid** [comment]
@yyb196 @shaloulcy any comments? 😄

**Comment by shaloulcy** [comment]
LGTM

**Pull request merged**